src/MartinGeorgiev/Doctrine/ORM/Query/AST/Functions/Cast.php

```diff
@@ -0,0 +1,31 @@
+<?php
+
+declare(strict_types=1);
+
+namespace MartinGeorgiev\Doctrine\ORM\Query\AST\Functions;
+
+use Doctrine\ORM\Query\SqlWalker;
+
+/**
+ * Implementation of PostgreSql CAST().
+ *
+ * @see https://www.postgresql.org/docs/current/sql-createcast.html
+ * @since 1.8.0
+ */
+class Cast extends BaseFunction
+{
+    protected function customiseFunction(): void
+    {
+        $this->setFunctionPrototype('cast(%s AS %s)');
+        $this->addNodeMapping('StringPrimary');
+        $this->addNodeMapping('StringPrimary');
+    }
+
+    public function getSql(SqlWalker $sqlWalker): string
+    {
+        $sql = parent::getSql($sqlWalker);
```

Comment:
Do you think the code-style can be reworked in the abstract way used in `IRegexp` and its corresponding test (`IRegexpTest`)?

Reply:
I've done that because we need something like: `CAST(column AS text)` and not `CAST(column AS 'text')`, we also can have something like: `CAST(column AS DECIMAL(8, 2))` (need to add a test for it).

Reply:
Hm, I see know what kind of a problem you intend to solve here. 

I am unaware of a nice-enough approach to achieve this with a node mapping, though. It feels like [Doctrine's documentation](https://www.doctrine-project.org/projects/doctrine-orm/en/latest/reference/dql-doctrine-query-language.html#document-syntax) doesn't have a relevant example.

I think there are unfomforatble security trade-offs with the use of `preg_replace` as the solution here. Instead, I suggest that the approach changes with dedicated, named functions that cast to the relevant datatype. For example, having separate classes (and registering those on an as-needed basis) from the likes of CAST_TO_JSON, CAST_TO_DECIMAL, CAST_TO_TEXT & etc.

Wdyt?

Reply:
The best way I think to avoid the regex, and keep it secure. Is to directly use the Lexer ? The example from Doctrine documentation, show that we need to use a trick to extract that value:

```php
        $this->type = $parser->getLexer()->token['value'];
```

An example of the implementation of CAST with Lexer: (From Beberlei extension):
```php
class Cast extends FunctionNode
{
    public Node $sourceType;

    public string $targetType;

    public function parse(Parser $parser): void
    {
        $parser->match(Lexer::T_IDENTIFIER);
        $parser->match(Lexer::T_OPEN_PARENTHESIS);

        $this->sourceType = $parser->SimpleArithmeticExpression();

        $parser->match(Lexer::T_AS);
        $parser->match(Lexer::T_IDENTIFIER);

        $type = $parser->getLexer()->token['value'];

        if ($parser->getLexer()->isNextToken(Lexer::T_OPEN_PARENTHESIS)) {
            $parser->match(Lexer::T_OPEN_PARENTHESIS);
            $parameter = $parser->Literal();
            $parameters = [$parameter->value];

            if ($parser->getLexer()->isNextToken(Lexer::T_COMMA)) {
                while ($parser->getLexer()->isNextToken(Lexer::T_COMMA)) {
                    $parser->match(Lexer::T_COMMA);
                    $parameter = $parser->Literal();
                    $parameters[] = $parameter->value;
                }
            }

            $parser->match(Lexer::T_CLOSE_PARENTHESIS);
            $type .= '('.implode(', ', $parameters).')';
        }

        $this->targetType = $type;

        $parser->match(Lexer::T_CLOSE_PARENTHESIS);
    }

    public function getSql(SqlWalker $sqlWalker): string
    {
        return sprintf(
            'CAST(%s AS %s)',
            $this->sourceType->dispatch($sqlWalker),
            $this->targetType
        );
    }
}

```

Reply:
I think this sounds like a better approach. Can you please add extensive test scenarios for this?

Reply:
@mpiot I addressed this thread in [bb82d3b](https://github.com/martin-georgiev/postgresql-for-doctrine/pull/116/commits/bb82d3ba158b4e1655e9ebdff4465a4a36ba8194). Hopefully, this is acceptable.